### ISSUE TITLE
Training with 3 codebooks

### ISSUE THREAD

**Issue by yonas-g** [hertz-pj/SNAC-Vocos#6]
Hi,

Thanks for providing the training code. I was wondering if it would be possible to train with 3 codebooks like [this one](https://huggingface.co/hubertsiuzdak/snac_24khz) by changing to the following configs.

```yaml
ratios: [5,4,2]
num_quantizers: 3
rand_num_quant: [4,4,4]
vq_scales: [4,2,1]
```

Thanks!

**Comment by w11wo** [comment]
Hi @yonas-g. I also wanted to do that. I found that this config was able to run:

```yaml
num_quantizers: 3
rand_num_quant: [4,4,4]
vq_scales: [4,2,1]
```

Changing the `ratios` parameter made it OOM all the time. I think that has more to do with the feature extractor's encoder, while the rest is related to the quantizer.

**Comment by hertz-pj** [comment]
Hi @yonas-g, as @w11wo  mentioned, the ratios determine the compression rate in the encoder section. The 5,4,2 configuration equates to 542=40x compression. With a sampling rate of 16,000, this would compress into 400 frames.

Regarding vq_scales: [4,2,1] means each VQ corresponds to 100, 50, and 25 frames respectively. This configuration should work fine for training.

Additionally, @w11wo, the OOM error is likely due to insufficient GPU memory caused by overly small ratios. You could try reducing the audio chunk length used for training. For example:
num_samples: 48,000 → 32,000/16,000

**Comment by yonas-g** [comment]
Thanks @hertz-pj,
I managed to train the the model with batch size of 2 but the loss became unstable after some steps as shown in the picture. Do you have any suggestions? Do you think lower batch size may have an impact?

<img width="1356" alt="Image" src="https://github.com/user-attachments/assets/4d73afbf-1ff1-4ebc-8d9c-303c1c181d6c" />

**Comment by yonas-g** [comment]
I kept `adanorm_num_embeddings` to 4. Do you think this has an issue?

**Comment by yonas-g** [comment]
@hertz-pj How do you handle the size miss match error during l1 error calculation? Do you suggesting taking the minimum length and cropping?

**Comment by Sedherthe** [comment]
@hertz-pj  @w11wo How much gpu memory is required to train this model?
